ee/elf-loader/src/elf.c

```diff
@@ -0,0 +1,85 @@
+/*
+# _____     ___ ____     ___ ____
+#  ____|   |    ____|   |        | |____|
+# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
+#-----------------------------------------------------------------------
+# (c) 20020 Francisco Javier Trujillo Mata <fjtrujy@gmail.com>
+# Licenced under Academic Free License version 2.0
+# Review ps2sdk README & LICENSE files for further details.
+*/
+
+#include <string.h>
+#include <sifrpc.h>
+#include <kernel.h>
+#include <sys/stat.h>
+#include <stdbool.h>
+#include <malloc.h>
+
+#include "elf.h"
+
+// Loader ELF variables
+extern u8 loader_elf[];
+extern int size_loader_elf;
+
+// ELF-loading stuff
+#define ELF_MAGIC 0x464c457f
+#define ELF_PT_LOAD 1
+
+int printf(const char *format, ...);
+
+static bool file_exists(const char *filename) {
+  struct stat   buffer;   
+  return (stat (filename, &buffer) == 0);
+}
+
+int LoadELFFromFile(const char *filename, int argc, char *argv[])
+{
+	u8 *boot_elf;
+	elf_header_t *eh;
+	elf_pheader_t *eph;
+	void *pdata;
+	int i;
+	int new_argc = argc + 1;
+	char **new_argv;
+	
+	// We need to check that the ELF file before continue
+	if (!file_exists(filename)) {
+		return -1; // ELF file doesn't exists
+	}
+	// ELF Exists
+	new_argv = malloc(new_argc);
+
+	new_argv[0] = (char *)filename;
+    for (i = 0; i < argc; i++) {
+		new_argv[i + 1] = argv[i];
+	}
```

Comment:
I'm sorry, but there's so many things wrong here:

1. Assuming "argc=1" and therefore "new_argc=2", you're allocating 2 bytes instead of 8 "malloc(new_argc*sizeof(char*));"
2. malloc also requires free, or you'll have a "memory leak". I can immagine in this specific case that it is intentional, but then you should comment why and how.
3. You're copying the POINTERS, not the DATA. This may seem to work fine in some situations, but will most likely fail in others. The pointers point to data from the OLD application. That data will be overwritten by the NEW application that is being loaded. To solve this you must copy this data to the loader's heap, assuming the heap is also below 0x100000, and not being overwritten by the new application.
4. And lastly, tabs or spaces, not both.

Reply:
Hi there!
1 and 2, I have changed back to static allocation.
3. Yes, you're right I'm copying just the pointers and not the content because I have seen how the `ExecPS2` function it-self creates new copies internally.
How I checked was printing the address memory of the `argvs parameters` in the `elf.c -> LoadELFFromFile` function and they were totally different than the one in the `loader.c -> main` function.
4. Thanks fixed!